### PR TITLE
docs: Refactor enhanced plural docs

### PR DESCRIPTION
**Pull request by amannn**
Hey @karimshalapy,

as discussed, here's a suggestion for how I'd restructure the pluralization docs.

While reworking the content, I think generally where things went in a slightly different direction, is that your proposed docs were very on the correct and elaborative side. I think it's very important to be correct and precise, but I feel like this sometimes gets in the way of making the docs easy to read and skim. My impression is that developers can quickly catch up something like `Hello {name}` and understand that `name` is a variable that can be interpolated without digging too much into the terminology and syntax.

I think the link to the ICU docs that you've included is quite important for those users who are interested in a deep dive, but I think I'd focus the `next-intl` docs a bit more on what we provide on top of the syntax, with expandable deep dives where there's a strong need for more clarity.

Therefore this proposal moves a few pieces back into place where they were before and adds supplemental deep dives for the pluralization features that were the original aim of this change.

Can you have a look and let me know what you think? I'd also really appreciate if you can confirm the statements about Arabic in the docs, they're based on a quick research on my side! 🙂 Thanks!

---

**Changes:**

 - Remove `danger` type for `Callout`
 - Remove invalid markers in `configuration`
 - Move plural and enums docs back to messages page and restructure content

### CODE REVIEW

File: docs/pages/docs/usage/messages.mdx [hunk 1]
```diff
@@ -157,84 +161,94 @@ t('message'); // "Hello world!"
...
 
-```js
-{
-  locales: ['en', 'ar'],
-  defaultLocale: 'en'
-}
-```
+For example, English has two forms: One for the singular (e.g. "1 _follower_") and one for everything else (e.g. "0 _followers_" and "2 _followers_"). In contrast, Chinese only has one form, while Arabic has six.
 
-Formatting might cause hydration errors because the locale strings that you have in your setup only contain the language with no extra data in it to help the [Intl API](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl) match the proper locale rules, so the Intl API resorts to a `localeMatcher` algorithm, and this algorirthm can result in different formatting rules in different hosting environments.
```
Comment (amannn) [inline docs/pages/docs/usage/messages.mdx:189]: The issue that we discussed in https://github.com/amannn/next-intl/issues/528 is really related to inconsistencies with the numbering system. This might apply to other languages too, but I don't have an overview of which languages might be affected. Also the issue might get resolved in the future if Chrome implements the fix.

Because of that, I think our discussion in the issue is currently the best documentation for the issue since it discusses a practical solution for Arabic for the current version of Chrome. I don't feel confident currently making broader suggestions about other languages and it's also somewhat hard keeping this up to date.

File: docs/pages/docs/usage/messages.mdx [hunk 2]
```diff
@@ -157,84 +161,94 @@ t('message'); // "Hello world!"
...
+
+```js
+t('message', {gender: 'female'}); // "She is online."
+```
+
+**Note**: The `other` case is required and will be used when none of the specific values match.
 
 ### Escaping
 
+Since curly braces are used for [interpolating dynamic values](/docs/usage/messages#interpolation-of-dynamic-values), you can escape them with the `'` marker to use the actual symbol in messages:
```
Comment (amannn) [inline docs/pages/docs/usage/messages.mdx:250]: Added an introductory description for every section on the page.

File: docs/pages/docs/usage/messages.mdx (hunk 1 above)
Comment (karimshalapy) [reply docs/pages/docs/usage/messages.mdx:189]: Regarding the different numbering systems, there are a lot of other languages that use different numbering systems. The first two that come to mind are Chinese and Thai.  
You can find a list of these languages [here](https://omniglot.com/language/numerals.htm).   

The issue most likely will exist for a very long period of time even if the Chrome dev team fixed this somehow, but the different server environments and all the other browsers will take a while to adapt and there are too many variables regarding this. That's why I thought it would be beneficial to mention that.  

That being said, I think that such an issue can be removed from the documentation and maybe reference the issue https://github.com/amannn/next-intl/issues/528 in the `README.md` or just leave it there for whoever searches the issues.

**Review by karimshalapy — COMMENTED** [review]
@amannn I like the changes that you made. They're concise and to the point. I love this [source](https://www.unicode.org/cldr/charts/43/supplemental/language_plural_rules.html). I didn't know it even exists 😄  

I was hesitant about adding other external links because you mentioned that you didn't want a lot of external links, but I thought about adding [this one](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules) myself.  

The only thing that I think can be enhanced is that it's not clear that there are different ways of formatting the messages interpolated variables out of the box, how to ignore this formatting when needed, and that the `#` is equal to `{variable, number}`. I found this pretty confusing at first but because I dug deeper I found out these stuff and I think they should be available in the documentation.  

If you don't find this valuable to have in the documentation somewhere, then I am good with the current state and we can merge it.

**Comment by karimshalapy** [comment]
@amannn 
TL;DR: Your statement is correct if you mean these 6 forms: `(0, 1, 2, few, many, other)` , although these should be called plural ranges because some of them share the same form of the word.


Regarding your request about the Arabic statement review.  
The only statement about the Arabic language that I found in the changes is the following.  

> For example, English has two forms: One for the singular (e.g. "1 _follower_") and one for everything else (e.g. "0 _followers_" and "2 _followers_"). In contrast, Chinese only has one form, while Arabic has six

It's a bit complicated, to be honest, so this statement is correct in a way.  
When it comes to cardinal pluralization, we have 5 forms of words:
- Singular
- Dual (I don't know how it translates to English but it's used when the count of something is 2)
- Sound Masculine Plurals
- Sound Feminine Plurals
- Broken Plurals

So, if you count the 0 state of the word we'll have 6 forms, but the 0 state shares the form with one of the plurals.  
It doesn't stop there, all these forms are not going to be used in this situation because the language is gendered, so you'll most likely at max use 3 forms of the word in the same sentence while interpolating plurals.  

Just to clarify something, by forms I mean how the word will look, and by state, I mean in which range of the count.  

So, if you're talking about how many forms of a word can be used in a sentence while interpolating plurals, then it should be 3.   

If you're talking about the different states `(0, 1, 2, few, many, other)` then they should be 6, but to be honest, the people using the `many` plural range will be a minority, as it's very niche because it uses the same as the Singular form but with an addition [Nunation](https://en.wikipedia.org/wiki/Nunation).  

If you're talking about all forms that can be used in the Arabic language then it should be 5 as well (the list mentioned above).  

And to clarify the difference between the state and form here's a simple table

| State | Form                                                          |
| ----- | ------------------------------------------------------------- |
| 0     | Singular                                                      |
| 1     | Singular                                                      |
| 2     | Dual                                                          |
| few   | Plural (broken, sound feminine, or sound masculine)             |
| many  | Singular (with Nunation, that can be ignored in Arabic Slang) |
| other | Singular                                                      |

I hope this was helpful, and if this wasn't the statement you wanted me to review, please let me know which one you meant.

**Comment by amannn** [comment]
> The only thing that I think can be enhanced is that it's not clear that there are different ways of formatting the messages interpolated variables out of the box, how to ignore this formatting when needed, and that the # is equal to {variable, number}. I found this pretty confusing at first but because I dug deeper I found out these stuff and I think they should be available in the documentation.

I tried to cover this with the note below the code snippets here:

Screenshot 2023-10-10 at 14 40 04

Do you think this is too concise?

I think `#` should really be used in this case. The formatting should not be ignored, as that will remove e.g. thousands and decimal separators. 

I think the solution in your case was really to use an explicit numbering system. We increasingly focus on the Server Components use case, where hydration mismatches should be a thing of the past. Therefore environment differences like the one you saw should hopefully be less and less applicable.



> It's a bit complicated, to be honest, so this statement is correct in a way.

Oh wow, thanks for including all the details! That's so interesting, I keep being amazed by all the nuances in languages. So to sum it up, I guess we can keep the statement, but there's some nuance to it? I got the idea from [the `Intl.PluralRules` docs on MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules) btw.

> I was hesitant about adding other external links because you mentioned that you didn't want a lot of external links, but I thought about adding [this one](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/PluralRules) myself.

I guess that was somewhat ambiguous, my apologies! I've now added documentation guidelines that hopefully provide some clarity for the future.

**Comment by karimshalapy** [comment]
@amannn You're totally correct, it should not be ignored, but knowing that `#` corresponds to `{variable, number}` is a valuable piece of information that I would miss without explicitly reading it somewhere because you might need to have an edge case where you'll use a custom format for this particular instance as follows: `{variable, number. <custom-format>}`. and without knowing that you can replace the `#` and use any other variable.  

This can happen for artistic reasons, the designer for some reason in a special case wants custom formatting for it, or whatever.

It's like a worthy mention that can be added in a collapsible menu just like you did with the plural ranges keywords.

That being said, I agree with you, and not a whole lot of people will face such scenarios, but if some face it they won't find the answer easily.

Regarding, the Arabic language statement, you can keep it, especially because you already linked: https://www.unicode.org/cldr/charts/43/supplemental/language_plural_rules.html 

I'll merge this PR for now, and if you think that the `#` is worthy of adding as a sidenote or a collapsable menu, it can be added later before merging my branch onto your main repo.
